app/controllers/courses_controller.rb

```diff
@@ -229,6 +229,109 @@ def users
             end
   end
 
+  action_auth_level :add_users_from_emails, :instructor
+  def add_users_from_emails
+    params.require([:user_emails, :role])
+    user_emails = params[:user_emails].split(/\n/).map(&:strip)
+
+    user_emails = user_emails.map do |email|
+      if email.nil?
+        nil?
+        # when it's first name <email>
+      elsif email =~ /(.*)\s+<(.*)>/
+        { first_name: Regexp.last_match(1), email: Regexp.last_match(2) }
+        # when it's first name last name <email>
+      elsif email =~ /(.*)\s+(.*)\s+<(.*)>/
+        { first_name: Regexp.last_match(1), last_name: Regexp.last_match(2),
+          email: Regexp.last_match(3) }
+        # when it's first name middle name last name <email>
+      elsif email =~ /(.*)\s+(.*)\s+(.*)\s+<(.*)>/
+        { first_name: Regexp.last_match(1), middle_name: Regexp.last_match(2),
+          last_name: Regexp.last_match(3), email: Regexp.last_match(4) }
+        # when it's email
+      else
+        { email: email }
+      end
+    end
+
+    # filter out nil emails
+    user_emails = user_emails.reject(&:nil?)
+
+    # check if email matches regex
+    email_regex = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
+
+    # raise error if any email is invalid and return which emails are invalid
+    invalid_emails = user_emails.reject { |user| user[:email] =~ email_regex }
+    if invalid_emails.any?
+      flash[:error] = "Invalid email(s): #{invalid_emails.map { |user| user[:email] }.join(', ')}"
```

Comment:
Is this for an additional layer of safety? Because I believe this error is already caught on the frontend, so a user would never see this.

Reply:
Yup exactly

Reply:
Generally I like to do validation on both sides. You don't want to assume what is calling the endpoint

Reply:
Yup, but then the testing plan of `Single entry: dave, assigned as instructor (error message of invalid email)` is wrong because the front end catches the error and states `Invalid input: dave` instead